### ISSUE TITLE
Core RETE doesn't implement algorithm correctly

### ISSUE THREAD

**Issue by woolfel** [NRules/NRules#306]
I was evaluating RETE engines for .Net and took some time to analyze the implementation. The design is fundamentally incorrect.  The join nodes (aka betanode) should have propagate assert left and assert right method. Instead the design has a generic propagateAssert(List). The reason this is important is the left and right inputs are fundamentally different. A betanode has 1 alphamemory and 1 betamemory.

You also need a specialized alphanode as the leftinput to the first betanode. In JESS, ernest called left-input-adapter node. This is a common design in CLIPS, JRules, JESS and other implementations. The design in the current main will work, but the performance with respect to ruleset size won't scale like RETE. The other impact is the runtime performance of the current design will be slower. I've helped other RETE implementations in the past and this is a common misunderstanding of RETE algorithm.

Basically, everyone makes this mistake the first time they try. It doesn't help that Dr. Forgy's paper doesn't describe these details, since it was obvious to him.

**Comment by snikolayev** [comment]
@woolfel sorry for the late response, I hope you are still interested in the subject. First of all, thanks for taking the time to look at the project and for providing your perspective. While I disagree with your assessment of the algorithm correctness, I want to make sure I properly understand your points and that I don't miss anything. At the same time, I want to clarify the design, so that it is easier to match the code you are looking at with the concepts defined in the Rete algorithm.
Regarding the join node - it does have left and right inputs, though they are not named *left* and *right*. They are differentiated by type signatures. There is `PropagateAssert` method that accepts facts, which corresponds to the right input, and there is `PropagateAssert` that accepts tuples, which corresponds to the left input. There are similar method pairs to propagate updates and retractions. In terms of terminology, what I'm calling *tuples* is what I believe is referred to as *tokens* in the original Dr Forgy's algorithm; or is also sometimes known as *WMEs lists*; but I prefer calling them tuples. Like you anticipated, there is a special left input node, serving as the first beta node, called `DummyNode`. This creates a chain of joins, starting from a zero-size tuple produced by the dummy node, incrementally joining facts from the right inputs, sourced by alpha memories. The last join in the chain produces an N-tuple that corresponds to the full rule match.
In that respect, I believe this implementation faithfully represents the idea of incremental construction of partial matches, and thus limits the scope of graph re-evaluation on fact inserts/updates/retracts, which is the essence of the Rete algorithm.
With that being said, NRules implementation has some performance improvements and algorithm extensions. One of which is that facts (and tuples) are not propagated one by one, but in lists, so that bulk changes to the working memory are processed more efficiently. So, perhaps it's those lists that looked unusual to you when comparing to the original Dr Forgy's implementation.
Please let me know if the above explanation makes sense to you and perhaps helps clarify the disconnect you encountered when comparing to other implementations. If you still believe there is a gap, and that I'm missing something in my explanation, please let me know and I'll be happy to discuss further.

**Comment by woolfel** [comment]
It might be good to add comments to your code, since the source for betaNode https://github.com/NRules/NRules/blob/develop/src/NRules/NRules/Rete/BetaNode.cs  has zero comments.  From the source code and comments, there's zero indication of the design. To be fair of the dozen RETE implementations I've read over the last 20 years most have very little comments. At least C# is easier to read than CLIPS C code, but that's mostly due to the fact I suck at C.

A couple of general suggestions on design for performance and ease of maintenance. The existential node https://github.com/NRules/NRules/blob/develop/src/NRules/NRules/Rete/ExistsNode.cs propagateAssert takes executionContext and stores the match count in the context. That works, but it's not optimal. I see the exectionContext class is quite simple and contains the session + working memory https://github.com/NRules/NRules/blob/develop/src/NRules/NRules/ExecutionContext.cs

A common approach used in my rule engine, JRules, CLILPS, JESS and many others is to have that in the working memory class instead. There's a few benefits to encapsulating that logic in the working memory. The first is you can have 1 RETE graph work with multiple working memories and context. When ernest added support sharing RETE network, he did something similar and refactored the design.

The "performance improvement" you describe is used by every mature RETE implementation, but most do not use List or collection. Instead arrays are used to avoid heap thrashing. Depending on RETE network size and dataset size, heap thrashing can slow down the execution by 2-10x.

I recommend reading the second version of Dr. Forgy's paper, which is more object oriented design than the original. It's also good to read his full Phd CS thesis, where he goes into greater detail. For example, the short paper doesn't describe specialized object type node that are directly attached to the rootNode.

hope that helps

**Comment by woolfel** [comment]
If you want to see how I implemented RETE, here's links to my engine.

Csharp version - _redacted_
Java version - _redacted_

I don't recommend people use it for real projects, since I don't want users. Feel free to learn from it or reuse anything you find useful. It's got some advanced features like network topology cost function, build-in MOLAP engine with support for snowflake schema, MDX like query for cubes, additional second order logic quantifiers, temporal distance calculation for memory management and a few other thiings.

**Comment by snikolayev** [comment]
@woolfel I again appreciate you taking the time to provide your perspective here. I assume you are trying to help, but unfortunately the way you are engaging is not helpful to me and is not leading to a constructive conversation that moves NRules project forward in a manner that's mindful of my time. You seem to have ignored my response, and just brought up more points that are factually inaccurate, that I now have to refute. While I will comment on those points below, I must ask you to follow the contributors guide and provide your feedback in a more structured way. If you see an issue with existing functionality, please provide a bug report with the specific repro scenario that I can work with. If you have a feature request, please provide a rationale and a proposed design and we can discuss.
I also don't want NRules project to be used as a platform to advertise your own implementation of a rules engine, as great as it might be. So, I will update your comment and redact the links. Please utilize other resources for promoting your own project.
Now to the specifics:
Regarding the comments in the code. As I indicated in the contributors' guide, I practice the comment-free coding style, where the code should be expressive enough to not need the comments. I don't want to have a philosophical argument about it, so this is something that contributors to this project have followed and so far, no one had a problem with.
Regarding storing existential matches in the working memory - that's how it works today, so please take a second look at the code. ExecutionContext simply carries a reference to the working memory, which is then used to store all session-level data.
And given that, the same Rete graph is shared between multiple sessions, as the graph itself is stateless, and all state is stored in the working memory.
Finally, a *List* in .net is a wrapper around an array, and if allocated with the expected capacity (which it is), is as efficient as a raw array, so again, your point is not helpful for the NRules implementation.
I'm going to close this thread at this point. If you feel like your request has not been answered adequately, please create a bug report or a feature request according to the guidelines provided in the contributors guide.
